### PR TITLE
Iot Endpoint Data Source

### PR DESCRIPTION
**Pull request by scottwinkler**
The iot endpoint is unique per account and is needed for connecting things to the iot registry. By having a data source for the iot endpoint, now it is possible to configure your things as part of a regular terraform deployment.

### CODE REVIEW

File: aws/data_source_aws_iot_endpoint_test.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+func TestAccAWSIotEndpoint(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccAWSIotEndpointConfig,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttrSet("aws_iot_endpoint.example", "id"),
```
Comment (bflad) [inline aws/data_source_aws_iot_endpoint_test.go]: The "resource" name here should be `data.aws_iot_endpoint.example`

```
=== RUN   TestAccAWSIotEndpoint
--- FAIL: TestAccAWSIotEndpoint (4.88s)
    testing.go:518: Step 0 error: Check failed: Check 1/2 error: Not found: aws_iot_endpoint.example in [root]
```

**Comment by scottwinkler** [comment]
Thanks, @bflad . I made the changes you requested. I am wondering if I should include the "endpoint-type" parameter for the data source. I could not find any documentation about what this type is, I think it might be something that will be used in the future but is not currently used. Should I include this parameter which doesn't appear to do anything to have a complete solution or leave this for a future enhancement? https://docs.aws.amazon.com/cli/latest/reference/iot/describe-endpoint.html

**Comment by bflad** [comment]
> Should I include this parameter which doesn't appear to do anything to have a complete solution or leave this for a future enhancement?

Interesting - its not well documented. I think we're okay leaving it out for now and can always add it in later.

File: aws/data_source_aws_iot_endpoint_test.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+func TestAccAWSIotEndpointDataSource(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccAWSIotEndpointConfig,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttrSet("data.aws_iot_endpoint.example", "id"),
```
Comment (bflad) [inline aws/data_source_aws_iot_endpoint_test.go]: Sorry I missed this the first review, we don't need to explicitly check `id`, its implicitly expected to be there as part of being a Terraform resource/data source. I'll remove this on merge 👍

File: website/docs/d/iot_endpoint.html.markdown [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+}
+```
+
+## Argument Reference
+
+N/A
+
+## Attributes Reference
+
+* `endpoint_address` - The endpoint. The format of the endpoint is as follows: identifier .iot.*region* .amazonaws.com.
```
Comment (bflad) [inline website/docs/d/iot_endpoint.html.markdown]: Formatting got a little wonky here, I'll fix on merge 👍

**Comment by bflad** [comment]
This has been released in [version 1.16.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1160-april-25-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**